File: code/programs/ruby/unix-tools/test/test_rm.rb
# frozen_string_literal: true

# test_rm.rb -- Tests for the Ruby rm tool
# ==========================================

require "simplecov"
SimpleCov.start do
  add_filter "/test/"
  enable_coverage :branch
  minimum_coverage 80
end

require "minitest/autorun"
require "tmpdir"
require "fileutils"
require "coding_adventures_cli_builder"

require_relative "../rm_tool"

require "stringio"

module RmTestHelper
  RM_TEST_SPEC = File.join(File.dirname(__FILE__), "..", "rm.json")

  def parse_rm_argv(argv)
    CodingAdventures::CliBuilder::Parser.new(RM_TEST_SPEC, ["rm"] + argv).parse
  end
end

class TestRmCliIntegration < Minitest::Test
  include RmTestHelper

  def test_basic_parse
    result = parse_rm_argv(["file.txt"])
    assert_kind_of CodingAdventures::CliBuilder::ParseResult, result
  end

  def test_help
    result = parse_rm_argv(["--help"])
    assert_kind_of CodingAdventures::CliBuilder::HelpResult, result
  end

  def test_version
    result = parse_rm_argv(["--version"])
    assert_kind_of CodingAdventures::CliBuilder::VersionResult, result
    assert_equal "1.0.0", result.version
  end

  def test_force_flag
    result = parse_rm_argv(["-f", "file.txt"])
    assert result.flags["force"]
  end

  def test_recursive_flag
    result = parse_rm_argv(["-r", "dir"])
    assert result.flags["recursive"]
  end

  def test_verbose_flag
    result = parse_rm_argv(["-v", "file.txt"])
    assert result.flags["verbose"]
  end

  def test_dir_flag
    result = parse_rm_argv(["-d", "dir"])
    assert result.flags["dir"]
  end
end

class TestRmConfirm < Minitest::Test
  def test_confirm_yes
    old_stdin = $stdin
    $stdin = StringIO.new("y\n")
    old_stderr = $stderr
    $stderr = StringIO.new
    assert rm_confirm("remove? ")
  ensure
    $stdin = old_stdin
    $stderr = old_stderr
  end

  def test_confirm_yes_full
    old_stdin = $stdin
    $stdin = StringIO.new("yes\n")
    old_stderr = $stderr
    $stderr = StringIO.new
    assert rm_confirm("remove? ")
  ensure
    $stdin = old_stdin
    $stderr = old_stderr
  end

  def test_confirm_no
    old_stdin = $stdin
    $stdin = StringIO.new("n\n")
    old_stderr = $stderr
    $stderr = StringIO.new
    refute rm_confirm("remove? ")
  ensure
    $stdin = old_stdin
    $stderr = old_stderr
  end

  def test_confirm_eof
    old_stdin = $stdin
    $stdin = StringIO.new("")
    old_stderr = $stderr
    $stderr = StringIO.new
    refute rm_confirm("remove? ")
  ensure
    $stdin = old_stdin
    $stderr = old_stderr
  end
end

class TestRmRemoveFile < Minitest::Test
  def test_remove_regular_file
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "file.txt")
      File.write(path, "content")
      assert rm_remove_file(path, force: false, interactive: false, recursive: false,
                            verbose: false, dir_flag: false, preserve_root: true)
      refute File.exist?(path)
    end
  end

  def test_remove_nonexistent_without_force
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "nonexistent")
      _out, err = capture_io do
        refute rm_remove_file(path, force: false, interactive: false, recursive: false,
                              verbose: false, dir_flag: false, preserve_root: true)
      end
      assert_includes err, "No such file or directory"
    end
  end

  def test_remove_nonexistent_with_force_succeeds
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "nonexistent")
      # With force=true, nonexistent files return true (silently succeed)
      assert rm_remove_file(path, force: true, interactive: false, recursive: false,
                            verbose: false, dir_flag: false, preserve_root: true)
    end
  end

  def test_remove_nonexistent_with_force_no_warning
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "nonexistent")
      _out, err = capture_io do
        rm_remove_file(path, force: true, interactive: false, recursive: false,
                       verbose: false, dir_flag: false, preserve_root: true)
      end
      refute_includes err, "No such file or directory"
    end
  end

  def test_remove_directory_without_recursive_or_dir
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "mydir")
      Dir.mkdir(path)
      _out, err = capture_io do
        refute rm_remove_file(path, force: false, interactive: false, recursive: false,
                              verbose: false, dir_flag: false, preserve_root: true)
      end
      assert_includes err, "Is a directory"
    end
  end

  def test_remove_directory_recursive
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "mydir")
      Dir.mkdir(path)
      File.write(File.join(path, "child.txt"), "data")
      assert rm_remove_file(path, force: false, interactive: false, recursive: true,
                            verbose: false, dir_flag: false, preserve_root: true)
      refute File.exist?(path)
    end
  end

  def test_remove_directory_recursive_verbose
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "mydir")
      Dir.mkdir(path)
      out, _err = capture_io do
        rm_remove_file(path, force: false, interactive: false, recursive: true,
                       verbose: true, dir_flag: false, preserve_root: true)
      end
      assert_includes out, "removed directory"
    end
  end

  def test_remove_empty_directory_with_dir_flag
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "emptydir")
      Dir.mkdir(path)
      assert rm_remove_file(path, force: false, interactive: false, recursive: false,
                            verbose: false, dir_flag: true, preserve_root: true)
      refute File.exist?(path)
    end
  end

  def test_remove_nonempty_directory_with_dir_flag
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "nonemptydir")
      Dir.mkdir(path)
      File.write(File.join(path, "child.txt"), "data")
      _out, err = capture_io do
        refute rm_remove_file(path, force: false, interactive: false, recursive: false,
                              verbose: false, dir_flag: true, preserve_root: true)
      end
      assert_includes err, "Directory not empty"
    end
  end

  def test_remove_dir_flag_verbose
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "emptydir2")
      Dir.mkdir(path)
      out, _err = capture_io do
        rm_remove_file(path, force: false, interactive: false, recursive: false,
                       verbose: true, dir_flag: true, preserve_root: true)
      end
      assert_includes out, "removed directory"
    end
  end

  def test_remove_file_verbose
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "file.txt")
      File.write(path, "data")
      out, _err = capture_io do
        rm_remove_file(path, force: false, interactive: false, recursive: false,
                       verbose: true, dir_flag: false, preserve_root: true)
      end
      assert_includes out, "removed '#{path}'"
    end
  end

  def test_preserve_root
    _out, err = capture_io do
      refute rm_remove_file("/", force: false, interactive: false, recursive: true,
                            verbose: false, dir_flag: false, preserve_root: true)
    end
    assert_includes err, "dangerous to operate"
  end

  def test_remove_symlink
    Dir.mktmpdir do |tmp|
      target = File.join(tmp, "target.txt")
      File.write(target, "data")
      link = File.join(tmp, "link.txt")
      File.symlink(target, link)
      assert rm_remove_file(link, force: false, interactive: false, recursive: false,
                            verbose: false, dir_flag: false, preserve_root: true)
      refute File.symlink?(link)
      assert File.exist?(target) # target still exists
    end
  end

  def test_interactive_confirm_yes
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "file.txt")
      File.write(path, "data")
      old_stdin = $stdin
      $stdin = StringIO.new("y\n")
      old_stderr = $stderr
      $stderr = StringIO.new
      result = rm_remove_file(path, force: false, interactive: true, recursive: false,
                              verbose: false, dir_flag: false, preserve_root: true)
      assert result
      refute File.exist?(path)
    ensure
      $stdin = old_stdin
      $stderr = old_stderr
    end
  end

  def test_interactive_confirm_no
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "file.txt")
      File.write(path, "data")
      old_stdin = $stdin
      $stdin = StringIO.new("n\n")
      old_stderr = $stderr
      $stderr = StringIO.new
      result = rm_remove_file(path, force: false, interactive: true, recursive: false,
                              verbose: false, dir_flag: false, preserve_root: true)
      # When user says "no", rm_confirm returns false, but the unless block
      # means the file ISN'T removed and we return true (skip).
      assert result
      assert File.exist?(path)
    ensure
      $stdin = old_stdin
      $stderr = old_stderr
    end
  end

  def test_interactive_recursive_directory
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "mydir")
      Dir.mkdir(path)
      old_stdin = $stdin
      $stdin = StringIO.new("y\n")
      old_stderr = $stderr
      $stderr = StringIO.new
      result = rm_remove_file(path, force: false, interactive: true, recursive: true,
                              verbose: false, dir_flag: false, preserve_root: true)
      assert result
      refute File.exist?(path)
    ensure
      $stdin = old_stdin
      $stderr = old_stderr
    end
  end

  def test_interactive_recursive_directory_decline
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "mydir")
      Dir.mkdir(path)
      old_stdin = $stdin
      $stdin = StringIO.new("n\n")
      old_stderr = $stderr
      $stderr = StringIO.new
      result = rm_remove_file(path, force: false, interactive: true, recursive: true,
                              verbose: false, dir_flag: false, preserve_root: true)
      assert result # returns true when declined
      assert File.exist?(path) # dir still exists
    ensure
      $stdin = old_stdin
      $stderr = old_stderr
    end
  end
end

class TestRmMainIntegration < Minitest::Test
  def test_main_removes_file
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "file.txt")
      File.write(path, "data")
      old_argv = ARGV.dup
      ARGV.replace([path])
      e = assert_raises(SystemExit) { rm_main }
      assert_equal 0, e.status
      refute File.exist?(path)
    ensure
      ARGV.replace(old_argv)
    end
  end

  def test_main_recursive
    Dir.mktmpdir do |tmp|
      dir = File.join(tmp, "mydir")
      Dir.mkdir(dir)
      File.write(File.join(dir, "child.txt"), "data")
      old_argv = ARGV.dup
      ARGV.replace(["-r", dir])
      e = assert_raises(SystemExit) { rm_main }
      assert_equal 0, e.status
      refute File.exist?(dir)
    ensure
      ARGV.replace(old_argv)
    end
  end

  def test_main_force_nonexistent
    old_argv = ARGV.dup
    ARGV.replace(["-f", "/nonexistent/file"])
    e = assert_raises(SystemExit) { rm_main }
    assert_equal 0, e.status
  ensure
    ARGV.replace(old_argv)
  end

  def test_main_help
    old_argv = ARGV.dup
    ARGV.replace(["--help"])
    out, _err = capture_io do
      e = assert_raises(SystemExit) { rm_main }
      assert_equal 0, e.status
    end
    assert_includes out, "rm"
  ensure
    ARGV.replace(old_argv)
  end

  def test_main_version
    old_argv = ARGV.dup
    ARGV.replace(["--version"])
    out, _err = capture_io do
      e = assert_raises(SystemExit) { rm_main }
      assert_equal 0, e.status
    end
    assert_includes out, "1.0.0"
  ensure
    ARGV.replace(old_argv)
  end

  def test_main_interactive_once_bulk
    Dir.mktmpdir do |tmp|
      files = (1..4).map do |i|
        p = File.join(tmp, "file#{i}.txt")
        File.write(p, "data")
        p
      end
      old_argv = ARGV.dup
      old_stdin = $stdin
      $stdin = StringIO.new("y\n")
      old_stderr = $stderr
      $stderr = StringIO.new
      ARGV.replace(["-I"] + files)
      e = assert_raises(SystemExit) { rm_main }
      assert_equal 0, e.status
    ensure
      ARGV.replace(old_argv)
      $stdin = old_stdin
      $stderr = old_stderr
    end
  end

  def test_main_interactive_once_decline
    Dir.mktmpdir do |tmp|
      files = (1..4).map do |i|
        p = File.join(tmp, "file#{i}.txt")
        File.write(p, "data")
        p
      end
      old_argv = ARGV.dup
      old_stdin = $stdin
      $stdin = StringIO.new("n\n")
      old_stderr = $stderr
      $stderr = StringIO.new
      ARGV.replace(["-I"] + files)
      e = assert_raises(SystemExit) { rm_main }
      assert_equal 0, e.status
      # Files should still exist since we declined
      files.each { |f| assert File.exist?(f) }
    ensure
      ARGV.replace(old_argv)
      $stdin = old_stdin
      $stderr = old_stderr
    end
  end

  def test_main_verbose
    Dir.mktmpdir do |tmp|
      path = File.join(tmp, "file.txt")
      File.write(path, "data")
      old_argv = ARGV.dup
      ARGV.replace(["-v", path])
      out, _err = capture_io do
        e = assert_raises(SystemExit) { rm_main }
        assert_equal 0, e.status
      end
      assert_includes out, "removed"
    ensure
      ARGV.replace(old_argv)
    end
  end

  def test_main_dir_fails_on_directory
    Dir.mktmpdir do |tmp|
      dir = File.join(tmp, "mydir")
      Dir.mkdir(dir)
      old_argv = ARGV.dup
      ARGV.replace([dir])
      _out, err = capture_io do
        e = assert_raises(SystemExit) { rm_main }
        assert_equal 1, e.status
      end
      assert_includes err, "Is a directory"
    ensure
      ARGV.replace(old_argv)
    end
  end
end

File: code/programs/python/unix-tools/tests/test_rm.py
"""Tests for the rm tool.

=== What These Tests Verify ===

These tests exercise the rm implementation, including:

1. Spec loading and CLI Builder integration
2. Removing files
3. The -r flag (recursive)
4. The -f flag (force)
5. The -v flag (verbose)
6. The -d flag (empty directories)
7. Safety checks (--preserve-root)
8. Business logic function (remove_file)
"""

from __future__ import annotations

import os
import sys
from pathlib import Path
from typing import Any

import pytest

SPEC_FILE = str(Path(__file__).parent.parent / "rm.json")

sys.path.insert(0, str(Path(__file__).parent.parent))

from rm_tool import remove_file


def parse_argv(argv: list[str]) -> Any:  # noqa: ANN401
    from cli_builder import Parser

    return Parser(SPEC_FILE, argv).parse()


class TestSpecLoading:
    def test_spec_file_exists(self) -> None:
        assert Path(SPEC_FILE).exists()

    def test_basic_parse(self) -> None:
        from cli_builder import ParseResult

        result = parse_argv(["rm", "file.txt"])
        assert isinstance(result, ParseResult)


class TestFlags:
    def test_force_flag(self) -> None:
        from cli_builder import ParseResult

        result = parse_argv(["rm", "-f", "file.txt"])
        assert isinstance(result, ParseResult)
        assert result.flags["force"] is True

    def test_recursive_flag(self) -> None:
        from cli_builder import ParseResult

        result = parse_argv(["rm", "-r", "dir"])
        assert isinstance(result, ParseResult)
        assert result.flags["recursive"] is True

    def test_verbose_flag(self) -> None:
        from cli_builder import ParseResult

        result = parse_argv(["rm", "-v", "file.txt"])
        assert isinstance(result, ParseResult)
        assert result.flags["verbose"] is True


class TestHelpVersion:
    def test_help(self) -> None:
        from cli_builder import HelpResult

        result = parse_argv(["rm", "--help"])
        assert isinstance(result, HelpResult)

    def test_version(self) -> None:
        from cli_builder import VersionResult

        result = parse_argv(["rm", "--version"])
        assert isinstance(result, VersionResult)
        assert result.version == "1.0.0"


class TestRemoveFile:
    def test_remove_regular_file(self, tmp_path: Path) -> None:
        filepath = str(tmp_path / "file.txt")
        Path(filepath).write_text("data")
        result = remove_file(
            filepath, force=False, interactive=False, recursive=False,
            verbose=False, dir_flag=False, preserve_root=True,
        )
        assert result is True
        assert not os.path.exists(filepath)

    def test_remove_nonexistent_fails(self, tmp_path: Path) -> None:
        filepath = str(tmp_path / "nonexistent")
        result = remove_file(
            filepath, force=False, interactive=False, recursive=False,
            verbose=False, dir_flag=False, preserve_root=True,
        )
        assert result is False

    def test_remove_nonexistent_force_ok(self, tmp_path: Path) -> None:
        filepath = str(tmp_path / "nonexistent")
        result = remove_file(
            filepath, force=True, interactive=False, recursive=False,
            verbose=False, dir_flag=False, preserve_root=True,
        )
        assert result is True

    def test_remove_directory_fails_without_r(self, tmp_path: Path) -> None:
        dirpath = str(tmp_path / "mydir")
        os.mkdir(dirpath)
        result = remove_file(
            dirpath, force=False, interactive=False, recursive=False,
            verbose=False, dir_flag=False, preserve_root=True,
        )
        assert result is False

    def test_remove_directory_recursive(self, tmp_path: Path) -> None:
        dirpath = tmp_path / "mydir"
        dirpath.mkdir()
        (dirpath / "file.txt").write_text("data")
        (dirpath / "subdir").mkdir()
        (dirpath / "subdir" / "nested.txt").write_text("nested")
        result = remove_file(
            str(dirpath), force=False, interactive=False, recursive=True,
            verbose=False, dir_flag=False, preserve_root=True,
        )
        assert result is True
        assert not dirpath.exists()

    def test_remove_empty_dir_with_d_flag(self, tmp_path: Path) -> None:
        dirpath = str(tmp_path / "emptydir")
        os.mkdir(dirpath)
        result = remove_file(
            dirpath, force=False, interactive=False, recursive=False,
            verbose=False, dir_flag=True, preserve_root=True,
        )
        assert result is True
        assert not os.path.exists(dirpath)

    def test_verbose_output(self, tmp_path: Path, capsys: pytest.CaptureFixture[str]) -> None:
        filepath = str(tmp_path / "verbose.txt")
        Path(filepath).write_text("data")
        remove_file(
            filepath, force=False, interactive=False, recursive=False,
            verbose=True, dir_flag=False, preserve_root=True,
        )
        captured = capsys.readouterr()
        assert "removed" in captured.out

    def test_preserve_root(self) -> None:
        result = remove_file(
            "/", force=False, interactive=False, recursive=True,
            verbose=False, dir_flag=False, preserve_root=True,
        )
        assert result is False

    def test_remove_symlink(self, tmp_path: Path) -> None:
        target = tmp_path / "target.txt"
        target.write_text("content")
        link = tmp_path / "link.txt"
        os.symlink(str(target), str(link))
        result = remove_file(
            str(link), force=False, interactive=False, recursive=False,
            verbose=False, dir_flag=False, preserve_root=True,
        )
        assert result is True
        assert not os.path.lexists(str(link))
        assert target.exists()  # Target should still exist.
